onnx/subbyte.py

```diff
@@ -70,3 +70,62 @@ def unpack_single_4bitx2(
     x_high = unpack_signed(x_high) if signed else x_high
     dtype = np.int8 if signed else np.uint8
     return (x_low.astype(dtype), x_high.astype(dtype))
+
+
+def float32_to_float4e2m1_unpacked(x: np.ndarray | np.dtype) -> np.ndarray:
+    """Cast to 4bit via rounding and clipping (without packing).
+
+    Args:
+        x: element to be converted
+
+    Returns:
+        An ndarray with a single float4e2m1 element (as uint8)
+    """
+
+    def float32_to_float4e2m1(value):
+        if np.isnan(value):
+            return 0x7
+        s = 0x0 if value >= 0 else 0x8
+        magnitude = np.abs(value)
+        if np.isinf(magnitude):
+            ret = 0x7
+        elif magnitude > 5:
+            ret = 0x7
+        elif magnitude >= 3.5:
+            ret = 0x6
+        elif magnitude > 2.5:
+            ret = 0x5
+        elif magnitude >= 1.75:
+            ret = 0x4
+        elif magnitude > 1.25:
+            ret = 0x3
+        elif magnitude >= 0.75:
+            ret = 0x2
+        elif magnitude > 0.25:
+            ret = 0x1
+        else:
+            ret = 0x0
+        return np.array(ret | s, dtype=np.uint8)
+
+    if not isinstance(x, np.ndarray):
+        x = np.asarray(x)
+    func = np.frompyfunc(float32_to_float4e2m1, 1, 1)
```

Comment:
Avoid frompyfunc, if possible?

Reply:
This one is actually a little tricky to rewrite as a numpy function because of the if block with many conditions. Any suggestions welcome.

Reply:
I would just create a mask for each one and fill in values into a preinitialized array incrementally.

Reply:
Added a new version.
Kept the old version as `float32_to_float4e2m1_unpacked_slow` but we could probably delete it. as it doesn't provide much reference value.

Reply:
I think we should remove it. As you said it doesn't provide much reference value.

Reply:
Removed.